src/compiler/scanner.rs

```diff
@@ -128,6 +157,63 @@ impl Scanner {
         token
     }
 
+    fn token_value(&self) -> &str {
+        self.token_value.as_ref().unwrap()
+    }
+
+    fn set_token_value(&mut self, token_value: &str) {
+        self.token_value = Some(token_value.to_string());
+    }
+
+    fn is_digit(&self, ch: char) -> bool {
+        ch >= CharacterCodes::_0 && ch <= CharacterCodes::_9
+    }
+
+    fn scan_number_fragment(&mut self) -> String {
+        let start = self.pos();
+        let result = "".to_string();
+        loop {
+            let ch = self.text().chars().nth(self.pos());
+            let ch = match ch {
+                Some(ch) => ch,
+                None => break,
+            };
+            if self.is_digit(ch) {
+                self.set_pos(self.pos() + 1);
+                continue;
+            }
+            break;
+        }
+        let mut ret = result;
+        ret.push_str(
+            &self
+                .text()
+                .chars()
+                .skip(start)
+                .take(self.pos() - start)
+                .collect::<String>(),
+        );
+        ret
+    }
+
+    fn scan_number(&mut self) -> ScanNumberReturn {
+        let start = self.pos();
+        let main_fragment = self.scan_number_fragment();
+        let end = self.pos();
+        let result: String = self.text().chars().skip(start).take(end - start).collect();
```

Comment:
I haven't done it, but it seems like you should be able to use the range operator for this rather than having to skip + take.

Reply:
This gets into the text-as-chars vs text-as-bytes thing, indexing a string with a range gives you a range of bytes rather than chars I believe